include mkl libraries for intel compiler on linux and macOS

iulusoy opened the pull request:
Add math library support with the new keyword `install_mkl`.
- include mkl libraries for all intel compiler versions on linux
- include mkl libraries for all ìntel-classic` versions on macOS EXCEPT version 2021.5
    - for some reason, installing the intel base-kit alongside the hpc-kit breaks the compiler set-up for version 2021.5
    - the exporting of `DYLD_LIBRARY_PATH` arbitrarily works on some runners and not on others, randomly fails for some version. If it is exported manually in the github action by setting it to `$MKLLIB`, the linking of the libraries works though. I could not test this much as the only macs I have for testing are the GitHub runners.

Initially I had wanted to only support linux. But then I tried to include `mkl` also for macOS - I consider this a starting point for future improvement. I am unfamiliar with windows, and the Fortran code I am working with has not been tested on Windows so far, so I did not include it.

Ideally, one would also want to install `BLAS` and `LAPACK` for the gnu compiler suite.

`mpi` support was also mentioned in #55

wpbonelli commented on setup-fortran.sh:
```diff
@@ -342,76 +380,128 @@ install_intel_apt()
+  fi
+
   source /opt/intel/oneapi/setvars.sh
+  # these exports have been removed in later versions of the action?
```
just moved to `install_intel` to de-duplicate here and in the windows function

wpbonelli commented on setup-fortran.sh:
```diff
@@ -342,76 +380,128 @@ install_intel_apt()
+      echo "ERROR: MACOS_BASEKIT_URL is empty - please check the version mapping for MKL"
+      echo "SKIPPING MKL installation..."
+    else
+      require_fetch
```
maybe worth a script to install intel on mac, like the one we have for windows

wpbonelli commented on setup-fortran.sh:
```diff
@@ -422,6 +512,18 @@ install_intel_dmg()
+  if $install_mkl; then
+    export MKLLIB="$ONEAPI_ROOT/mkl/latest/lib"
+    export MKLROOT="$ONEAPI_ROOT/mkl/latest"
+    export DYLD_LIBRARY_PATH="$MKLLIB":$DYLD_LIBRARY_PATH
```
I think we probably shouldn't edit `DYLD_LIBRARY_PATH` on principle

wpbonelli commented:
thanks for this @iulusoy- I think if we can also support windows, we might as well include mkl by default and not have a separate input? it might be tricky to find the oneAPI basekit links for older versions, it is unfortunate that's not in the hpckit

iulusoy commented:
Unfortunately I have no experience with windows, I can try but it will have to be in three-four weeks.

iulusoy commented:
Incorporated the requested changes so far, could you please re-review @wpbonelli ?
Windows not incorporated yet.

wpbonelli commented on install-mkl-macos.sh:
```diff
@@ -0,0 +1,22 @@
+MACOS_BASEKIT_URL=$1
```
I think we can make this script generic like `install-intel-windows.bat`, so we can pass the URL and components as args and call it for both basekit and hpckit?

wpbonelli commented on .github/compat/long_compat.csv:
```diff
@@ -1,4 +1,5 @@
 runner,compiler,version,support
+macos-13,intel-classic,2021.5,&check;runner,compiler,version,support
```
Only the first line of this file should have headers.

In general when we add support for new toolchains/versions we need to

* update `matrix.yml` with new version combinations we expect to support
* manually trigger the CI in reporting mode (on your PR branch on your fork)
* get the diffs for `long_compat.csv`, `compat.csv`, and `README.md`, and splice them into the PR
* make sure CI passes again in test mode before the PR merges

Maybe we should add a developer doc with info on this workflow, I can do that.

iulusoy replied:
Documentation on this would be great! How should mkl be included, in the same table or in a separate one? Could be a bit too much information at once if it's in the same table.

wpbonelli commented on .github/actions/test-fc/action.yml:
```diff
@@ -58,6 +58,22 @@ runs:
         }
         rm hw.exe
 
+    - name: Test compile Fortran (pwsh)
```
looks like accidental duplication of lines 46-59?

wpbonelli commented on test/hw_mkl.f90:
```diff
@@ -0,0 +1,13 @@
+program hello
```
It's good to include math ops in the test program. If we include mkl by default, we can just roll this into the existing fortran test program/action.

iulusoy replied:
Yes, but that would also require installing BLAS/LAPACK for the gnu compilers - for now I would leave it separate, and make the optional math library install the default version once there is something for all compilers/OS.

wpbonelli commented on action.yml:
```diff
@@ -129,6 +135,7 @@ runs:
         echo FC=$FC>>$GITHUB_ENV
         echo CC=$CC>>$GITHUB_ENV
         echo CXX=$CXX>>$GITHUB_ENV
+        echo "MKLLIB=$MKLLIB" >> $GITHUB_ENV
```
if `MKLLIB` and `MKLROOT` are set by the oneAPI activation script, the `export_intel_vars` function in `setup-fortran.sh` should handle them, no need to set separately here.

However the previous step may need modification to support cache restore on Windows, `set | findstr /c:"oneAPI" >> "%GITHUB_ENV%"` may not cover everything we need

wpbonelli commented:
thanks @iulusoy. I might take a shot at Windows if I have a chance in the meantime

iulusoy commented:
Addressed the comments, imo opinion the next steps would be (if math libraries are to be installed as default):
- include mkl install for Windows
- include BLAS/LAPACK install for gnu compilers: Linux, MacOS, Windows
I would suggest doing those in separate PRs to minimize complexity.

wpbonelli commented:
@iulusoy yes, we'd also want to look at nvidia and lfortran too. Agreed on keeping PRs small. I think a lot of people use `@main` instead of the tags so I think it's best to wait for all platforms/toolchains before merging here. I made a [`math`](https://github.com/fortran-lang/setup-fortran/tree/math) branch we can use for prototyping, if you want to move this PR over there? I will make a few more quick comments in the meantime
